server/src/test/java/org/elasticsearch/script/field/WriteFieldTests.java

```diff
@@ -557,11 +619,14 @@ public void testHasValue() {
     public void testGetIndex() {
         Map<String, Object> root = new HashMap<>();
         root.put("a", Map.of("b", List.of(1, 2, 3, 5), "c", "foo"));
-        WriteField wf = new WriteField("a.b", () -> root);
-        assertEquals(5, wf.get(3, 100));
+        assertEquals(5, new WriteField("a.b", () -> root).get(3, 100));
+        assertEquals(5, new WriteField("a.b.3", () -> root).get(100));
         assertEquals(100, new WriteField("c.d", () -> root).get(3, 100));
+        assertEquals(100, new WriteField("c.d.3", () -> root).get(100));
         assertEquals("bar", new WriteField("a.c", () -> root).get(1, "bar"));
+        assertEquals("bar", new WriteField("a.c.1", () -> root).get("bar"));
         assertEquals("foo", new WriteField("a.c", () -> root).get(0, "bar"));
+        assertEquals("bar", new WriteField("a.c.0", () -> root).get("bar"));
```

Comment:
This is a special case where `field('a.c').get(0, 'bar')` will return a different value compared to `field('a.c.0').get('bar')`. That's because the former will return a result if the value is a scalar (in this case `c` has the scalar value `"foo"`), whereas the latter either requires `c` to be a list that has at least one element or that `c` is a map that has a key named `0`. Because of that ambiguity of whether `0` is an index or a key in the map, we don't have the exact same behavior across the two different ways to access that field. Probably not a big deal but worth pointing out.